src/sidebar/services/autosave.js

```diff
@@ -0,0 +1,79 @@
+/**
+ * A service for automatically saving new highlights.
+ */
+const MAX_RETRIES = 3;
+
+// @ngInject
+export default function autosaveService(annotationsService, store) {
+  // A Set of annotation $tags that have save requests in-flight
+  const saving = new Set();
+
+  // Map annotation $tags to number of save attempts
+  const retries = new Map();
+
+  /**
+   * Determine whether we should try to send a save request for the highlight
+   * indicated by `htag`
+   *
+   * @param {string} htag - The local unique identifier for the unsaved highlight
+   * @return {boolean}
+   */
+  const shouldSaveHighlight = htag => {
+    if (saving.has(htag)) {
+      // We're already saving it
+      return false;
+    }
+    if (retries.has(htag) && retries.get(htag) >= MAX_RETRIES) {
+      // We've tried too many times already
+      return false;
+    }
+    return true;
+  };
+
+  /**
+   * Store-subscribed call back. Automatically save new highlights.
+   */
+  const autosaveNewHighlights = () => {
+    const newHighlights = store.newHighlights() || [];
+
+    newHighlights.forEach(highlight => {
+      // Because this is a new annotation object, it does not yet have an `id`
+      // property. Use the local `$tag` for uniqueness instead.
+      const htag = highlight.$tag;
+
+      if (!shouldSaveHighlight(htag)) {
+        return;
+      }
+
+      if (!retries.has(htag)) {
+        // This is the first time we've tried to save this highlight. Keep track
+        // of that.
+        retries.set(htag, 0);
+      }
+
+      saving.add(htag);
+
+      annotationsService
+        .save(highlight)
+        .then(() => {
+          // We've been successful, so we can remove the tracking of retries
+          retries.delete(htag);
+        })
+        .catch(() => {
+          // We've failed, so increment try count
+          let tryCount = retries.get(htag);
+          retries.set(htag, tryCount++);
```

Comment:
This code isn't doing what you intended!

```
$ node
Welcome to Node.js v13.8.0.
Type ".help" for more information.
> var m = new Map
undefined
> var i = 1
undefined
> m.set('foo', i++)
Map(1) { 'foo' => 1 }
> m.get('foo')
1
>
```

Reply:
DRAT IT! This code is out-of-date somehow. I fixed the increment to be prefix locally, because, yeah